File: services/balance/src/connect.ts
import { ConnectRouter } from "@bufbuild/connect";
import { IntegrationService } from "@shumkovdenis/protobuf-schema/lib/integration/v1/api_connect";

export default (router: ConnectRouter) =>
  router.service(IntegrationService, {
    async getBalance(req, ctx) {
      console.log("req:traceparent", ctx.requestHeader.get("traceparent"));
      console.log("req:grpc-trace-bin", ctx.requestHeader.get("grpc-trace-bin"));
      console.log("res:traceparent", ctx.responseHeader.get("traceparent"));
      console.log("res:grpc-trace-bin", ctx.responseHeader.get("grpc-trace-bin"));
      return { balance: BigInt(1010) };
    }
  });

File: services/remote/connect.go
package main

import (
	"context"
	"fmt"
	"net/http"

	"github.com/bufbuild/connect-go"
	integration "github.com/shumkovdenis/protobuf-schema/gen/integration/v1"
	integrationConnect "github.com/shumkovdenis/protobuf-schema/gen/integration/v1/integrationv1connect"
	"github.com/shumkovdenis/services/remote/helpers"
	"golang.org/x/net/http2"
	"golang.org/x/net/http2/h2c"
)

type ConnectServer struct {
	integrationConnect.UnimplementedIntegrationServiceHandler
}

func NewConnectServer(cfg Config) error {
	var server ConnectServer

	mux := http.NewServeMux()
	mux.Handle(integrationConnect.NewIntegrationServiceHandler(
		&server,
		connect.WithInterceptors(
			helpers.NewTraceInterceptor(cfg.GRPCTrace),
			helpers.NewLoggerInterceptor(),
		),
	))

	return http.ListenAndServe(
		fmt.Sprintf(":%d", cfg.Port),
		// Use h2c so we can serve HTTP/2 without TLS.
		h2c.NewHandler(mux, &http2.Server{}),
	)
}

func (s *ConnectServer) GetBalance(
	ctx context.Context,
	req *connect.Request[integration.GetBalanceRequest],
) (*connect.Response[integration.GetBalanceResponse], error) {
	res := connect.NewResponse(&integration.GetBalanceResponse{
		Balance: 9999,
	})
	return res, nil
}
